openshift/default-catalog-consistency/pkg/check/filesystem.go

```diff
@@ -38,3 +41,78 @@ func FilesystemHasDirectories(paths ...string) FilesystemCheck {
 		},
 	}
 }
+
+// FilesystemHasExecutables checks whether the specified paths refer to existing, accessible, and executable file.
+// It handles both regular files and symbolic links:
+//   - For regular files: checks that the file exists and is executable.
+//   - For symlinks: follows the link and checks that the target is an executable regular file.
+//
+// Any issues (missing file, broken symlink, non-executable file, etc.) will be reported as errors.
+func FilesystemHasExecutables(root string, paths ...string) FilesystemCheck {
+	return FilesystemCheck{
+		Name: "FilesystemHasExecutables" + fmt.Sprintf(":(%q)", paths),
+
+		Fn: func(_ context.Context, _ fs.FS) error {
+			var errs []error
+
+			// ExecPerm useful to check the executable permission
+			// 0o100 (owner), 0o010 (group), 0o001 (any)
+			const ExecPerm = 0o111
+
+			for _, rel := range paths {
+				full := filepath.Join(root, rel)
+
+				fi, err := os.Lstat(full)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("not found %q: %w",
+						rel, err))
+					continue
+				}
+
+				switch mode := fi.Mode(); {
+				// Case 1: The path is a binary/executable
+				case mode.IsRegular():
+					// Check if any executable permissions are set; if not, report as non-executable
+					if mode.Perm()&ExecPerm == 0 {
+						errs = append(errs, fmt.Errorf("%q is a file but not executable", rel))
+					}
+
+				// Case 2: The path is a symbolic link for the binary/exacutable
+				case mode&os.ModeSymlink != 0:
+					// Resolve the symlink to its target path
+					target, err := os.Readlink(full)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("unable to reach symlink %q: %w", rel, err))
+						continue
+					}
+
+					if !filepath.IsAbs(target) {
+						target = filepath.Join(filepath.Dir(full), target)
+					}
+
+					// Get the info of the binary/target
+					tgtInfo, err := os.Stat(target)
+					if err != nil {
+						errs = append(errs,
+							fmt.Errorf("find symlink %q which points to the target %q that doesn't exist: %w",
```

Comment:
Same as LStat, this is probably too specific about the reason, could be permissions, etc,

Reply:
👍 Updated thank you.